README.md
# PROJECTS

#### MATHSTER

#### OMICRON NEWS PRINTER




News_Parser/news.txt
[{'source': {'id': None, 'name': 'New York Times'}, 'author': 'Sharon LaFraniere', 'title': 'Revised Moderna Vaccine Works Better Against Omicron, Trial Suggests', 'description': 'The findings indicate that the updated vaccine produced a significantly stronger immune response against Omicron, but it is unclear how it will fare against future versions of the virus.', 'url': 'https://www.nytimes.com/2022/06/08/us/politics/moderna-vaccine-omicron-variant.html', 'urlToImage': 'https://static01.nyt.com/images/2022/06/08/us/politics/08virus-briefing-moderna/08virus-briefing-moderna-facebookJumbo.jpg', 'publishedAt': '2022-06-08T12:18:24Z', 'content': 'But Omicron has been spawning subvariants for months, and some vaccine experts say what matters now is how well a new booster formulation would protect against the latest subvariants, BA.4 and BA.5, … [+1644 chars]'}, {'source': {'id': None, 'name': 'New York Times'}, 'author': 'Sharon LaFraniere', 'title': 'The F.D.A. authorizes Pfizer-BioNTech boosters for children ages 5 to 11.', 'description': 'More than eight million of the 28 million children in that age group in the United States have received two vaccine shots.', 'url': 'https://www.nytimes.com/2022/05/17/us/politics/fda-pfizer-booster-children.html', 'urlToImage': 'https://static01.nyt.com/images/2022/05/17/world/17virus-briefing-fda-pfizer1/merlin_197274552_b9493f17-f75d-4047-bee9-cab6a523c524-facebookJumbo.jpg', 'publishedAt': '2022-05-17T14:50:40Z', 'content': 'In a statement, Dr. Robert M. Califf, the F.D.A.s commissioner, said: While it has largely been the case that Covid-19 tends to be less severe in children than adults, the Omicron wave has seen more … [+1278 chars]'}, {'source': {'id': None, 'name': 'Gizmodo.com'}, 'author': 'Lauren Leffer', 'title': 'Covid Boosters Authorized for Kids as Young as 5', 'description': 'On Tuesday morning, the FDA announced that it was officially authorizing the use of covid-19 booster shots for children 5-11 years old. Specifically, the administration is authorizing a single, additional Pfizer-BioNTech booster dose for kids at least five mo…', 'url': 'https://gizmodo.com/fda-authorizes-pfizer-booster-shots-covid-19-kids-ages-1848936534', 'urlToImage': 'https://i.kinja-img.com/gawker-media/image/upload/c_fill,f_auto,fl_progressive,g_center,h_675,pg_1,q_80,w_1200/d8fac7212c47e8697be83f1fa23dbbd7.jpg', 'publishedAt': '2022-05-17T16:20:00Z', 'content': 'On Tuesday morning, the FDA announced that it was officially authorizing the use of covid-19 booster shots for children 5-11 years old. Specifically, the administration is authorizing a single, addit… [+2299 chars]'}, {'source': {'id': None, 'name': 'The Guardian'}, 'author': 'Ian Sample Science editor', 'title': 'Trials of new Covid vaccine raise hopes of once-a-year booster', 'description': 'Moderna shot is said to produce eightfold increase in antibody levels against Omicron in first four weeksAn updated version of Moderna’s Covid vaccine produces an eightfold increase in antibody levels against the Omicron variant, according to early trial resu…', 'url': 'https://amp.theguardian.com/world/2022/jun/08/trials-of-new-covid-vaccine-raise-hopes-of-once-a-year-booster', 'urlToImage': 'https://i.guim.co.uk/img/media/cc0b50b58de0cb34efb3f5fd9756dd15821a8638/0_61_3740_2244/master/3740.jpg?width=1200&height=630&quality=85&auto=format&fit=crop&overlay-align=bottom%2Cleft&overlay-width=100p&overlay-base64=L2ltZy9zdGF0aWMvb3ZlcmxheXMvdGctZGVmYXVsdC5wbmc&enable=upscale&s=0c8cba7d1245bfab9985a8f5fb1fc0be', 'publishedAt': '2022-06-08T17:55:12Z', 'content': 'An updated version of Modernas Covid vaccine produces an eightfold increase in antibody levels against the Omicron variant, according to early trial results, raising hopes for a once-a-year booster t… [+3523 chars]'}, {'source': {'id': None, 'name': 'New York Times'}, 'author': 'The New York Times', 'title': 'Relief and Reunions After Shanghai Reopens, but Also Anxieties', 'description': 'The authorities kept some restrictions, such as requiring residents to show negative Covid test results to use the subway and limiting access to housing compounds.', 'url': 'https://www.nytimes.com/live/2022/06/01/world/covid-19-mandates-vaccine-cases', 'urlToImage': 'https://static01.nyt.com/images/2022/06/01/world/01virus-china-shanghai1/01virus-china-shanghai1-facebookJumbo.jpg', 'publishedAt': '2022-06-01T09:21:31Z', 'content': 'Despite strong levels of vaccination among older people, Covid killed them at vastly higher rates during this winters Omicron wave than it did last year, preying on long delays since their last shots… [+12143 chars]'}, {'source': {'id': 'reuters', 'name': 'Reuters'}, 'author': None, 'title': "Omicron sub-variants driving Africa's COVID surge - WHO - Reuters.com", 'description': 'Sub-variants of the Omicron variant of COVID-19 and a relaxation of public health measures are driving a current surge in infections in southern Africa, a senior official from the World Health Organization (WHO) said on Thursday.', 'url': 'https://www.reuters.com/world/africa/omicron-sub-variants-driving-africas-covid-surge-who-2022-05-12/', 'urlToImage': 'https://www.reuters.com/resizer/W9EfC-kpU1Z1ecaTeIDIEv7gxzc=/1200x628/smart/filters:quality(80)/cloudfront-us-east-2.images.arcpublishing.com/reuters/KLEV25WQDZMDFPQIRH47PBBIG4.jpg', 'publishedAt': '2022-05-12T10:29:00Z', 'content': 'JOHANNESBURG, May 12 (Reuters) - Sub-variants of the Omicron variant of COVID-19 and a relaxation of public health measures are driving a current surge in infections in southern Africa, a senior offi… [+606 chars]'}, {'source': {'id': 'reuters', 'name': 'Reuters'}, 'author': None, 'title': 'Britain reduce COVID alert level as Omicron subsides - Reuters UK', 'description': 'Britain reduced its COVID-19 alert level from four to three on Friday, saying the Omicron-variant wave of the virus was subsiding and healthcare pressures continued to decrease in all nations.', 'url': 'https://www.reuters

News_Parser/test.py
from fpdf import FPDF

pdf = FPDF()
pdf.add_page()
pdf.set_font('helvetica', size=12)
pdf.cell(txt="hello world")
pdf.output("hello_world.pdf")

Mathster Project/test.py
from project import main

main()

Mathster Project/test_project.py
from project import score, ranker, next_rank, welcome, user, level_1
import csv
import pytest

# tester1,10
# test,10
# troy,30
# roy,120


def test_score():
    assert score("roy") == int(case_1)
    assert score("troy") == int(case_2)
    assert score("test") == int(case_3)
    assert score("tester1") == int(case_4)
    with pytest.raises(IndexError):
        # looks up a non-existent user, expects IndexError
        score("sadjas@3232!")


"""
returns rank according to score
>90 - "Math_Elite"
80-89 - "Math_Diamond"
60-79 - "Math_Platinum"
50-59 - "Math_Gold"
40-49 - "Math_Silver+"
30-39 - "Math_Silver"
20-29 - "Math_Bronze+"
10-19 - "Math_Bronze"
"""
def test_ranker():
    assert ranker(50) == ("Math_Gold", 50)
    assert ranker(3) == ("Noob", 3)
    assert ranker(89) == ("Math_Diamond", 89)
    assert ranker(1337) == ("Math_Elite", 1337)
    assert ranker(0) == ("Noob", 0)
    # tests for non-integer
    with pytest.raises(TypeError):
        assert ranker('cat')

    # tests for non-negative integer
    with pytest.raises(ValueError):
        assert ranker(-1)

def test_next_rank():
    assert next_rank(('Math_Gold', 50)) == ("Math_Gold", 50)
    assert next_rank(('Math_Elite', 99)) == ("Math_Elite", 99)
    with pytest.raises(ValueError):
        assert next_rank('nonsense')

def test_welcome():
    assert welcome() == None

def test_user():
    with pytest.raises(OSError):
        assert user()

def test_level_1():
    with pytest.raises(OSError):
        assert level_1('roy')


def main():
    test_score()
    test_ranker()
    test_next_rank()
    test_welcome()
    test_user()
    test_level_1()

with open("scores.csv") as f:
    columns = ["name", "score"]
    reader = csv.DictReader(f, columns)
    for row in reader:
        name_key = row["name"]
        score_key = row["score"]
        if name_key == "roy":
            case_1 = score_key
        elif name_key == "troy":
            case_2 = score_key
        elif name_key == "test":
            case_3 = score_key
        elif name_key == "tester1":
            case_4 = score_key


project/level1.py
# level1.py
# contains list of dictionary of questions

import sys
import csv

class Level_One:
    def __init__(self, name, questions=10):
        self.questions = questions
        self.counter = 0 # score of player
        self.name = name


    def __str__(self):
        print(f"Hi, {self.name}. There are a total of {self.questions} questions. Do your best!")
        return (f"Each question is worth 1 mark. Current score is {self.counter}.")

    # # deprecated
    # @classmethod
    # def get(cls):
    #     name = input("Choose a player name: ").lower()
    #     return cls(name)

    @property
    def questions(self):
        return self._questions

    @questions.setter
    def questions(self, questions):
        try:
            if questions == 0:
                if self.counter == 10:
                    self.save_score()
                    print(f"Congratulations, {self.name}. Perfect score of {self.counter}! Completed level one.")
                    return
                else:
                    self.save_score()
                    print(f"Congratulations, {self.name}. Completed level one with a score of {self.counter}. Keep it up!")
                    return
            self._questions = questions
        except IndexError:
            with open("scores.csv", "a") as f:
                f.seek(0)
                columns = ["name", "score"]
                writer = csv.DictWriter(f, columns)
                writer.writerow({"name": self.name, "score": self.counter})
                print(f"Congratulations, {self.name}. Completed level one with a score of {self.counter}. Keep it up!")
                return
    @property
    def counter(self):
        return self._counter

    @counter.setter
    def counter(self, counter):
        self._counter = counter

    def check_ready(self):
        while True:
            status = input("Ready? Y/N\n")
            if status.lower() == "y":
                break
            elif status.lower() == "n":
                continue
            else:
                print("Invalid input. Try again.")
                continue

    def save_score(self):
        with open("scores.csv", "r+") as f:
            columns = ["name", "score"]
            reader = csv.DictReader(f, columns)
            # creates a list of dictionaries with name not matching player name
            filtered_output = [line for line in reader if line["name"] != self.name]

            # resets read position of csv file
            f.seek(0)
            # creates a list of dictionary with name matching player name to store score key-value pair
            past_record = [line for line in reader if line["name"] == self.name]
            # indexes into list, using score as a key to obtain its valuem, then adding to current score
            score = int((past_record[0])["score"]) + self.counter
            # dictionary
            update = {"name": self.name, "score": score}
            filtered_output.append(update)

            # resets read position of csv file
            f.seek(0)
            writer = csv.DictWriter(f, columns)
            # rewrites the file
            writer.writerows(filtered_output)
            return score


    def q1(self):
        question = input("If 1/10 is 0.1, what is 2/10 in decimal?\n")
        solution = "0.2"
        if question == solution:
            self.counter += 1
            print("Current score:", self.counter)
            self.questions -= 1
        else:
            i = 1
            while i < 4:
                print(f"Hint: The numerator matches the tenth place for fractions with 10 as denominator. Try again!\nAttempt {i}: ", end="")
                attempt = input()
                if attempt == solution:
                    self.counter += 1
                    print("Current score:", self.counter)
                    self.questions -= 1
                    return
                else:
                    i += 1
            print("The correct answer is 0.2. Try better next time!")
            self.questions -= 1

    def q2(self):
        question = input("If 5/10 is 0.5, what is 50/100 in decimal?\n")
        solution = "0.5"
        if question == solution:
            self.counter += 1
            print("Current score:", self.counter)
            self.questions -= 1
        else:
            i = 1
            while i < 4:
                print(f"Hint: The numerator matches the tenth place for fractions with 10 as denominator. Try again!\nAttempt {i}: ", end="")
                attempt = input()
                if attempt == solution:
                    self.counter += 1
                    print("Current score:", self.counter)
                    self.questions -= 1
                    return
                else:
                    i += 1
            print("The correct answer is 0.5. Try better next time!")
            self.questions -= 1


    def q3(self):
        question = input("If 23/100 is 0.23, what is 46/100 in decimal?\n")
        solution = "0.46"
        if question == solution:
            self.counter += 1
            print("Current score:", self.counter)
            self.questions -= 1
        else:
            i = 1
            while i < 4:
                print(f"Hint: What does the numerator and the tenth & hundredth place have in common? Try again!\nAttempt {i}: ", end="")
                attempt = input()
                if attempt == solution:
                    self.counter += 1
                    print("Current score:", self.counter)
                    self.questions -= 1
                    return
                else:
                    i += 1
            print("The correct answer is 0.46. Try better next time!")
            self.questions -= 1

    def q4(self):
        question = input("If 888/1000 is 0.888, what is 463/1000 in decimal?\n")
        solution = "0.463"
        if question == solution:
            self.counter += 1
            print("Current score:", self.counter)
            self.questions -= 1
        else:
            i = 1
            while i < 4:
                print(f"Hint: What does the numerator and the tenth, hundredth & thousandth place have in common? Try again!\nAttempt {i}: ", end="")
                attempt = input()
                if attempt == solution:
                    self.counter += 1
                    print("Current score:", self.counter)
                    self.questions -= 1
                    return
                else:
                    i += 1
            print("The correct answer is 0.463. Try better next time!")
            self.questions -= 1

    def q5(self):
        question = input("What is 3 divided by 4 in fraction?\n")
        solution = "3/4"
        if question == solution:
            self.counter += 1
            print("Current score:", self.counter)
            self.questions -= 1
        else:
            i = 1
            while i < 4:
                print(f"Hint: Try writing it out using the division sign. Try again!\nAttempt {i}: ", end="")
                attempt = input()
                if attempt == solution:
                    self.counter += 1
                    print("Current score:", self.counter)
                    self.questions -= 1
                    return
                else:
                    i += 1
            print("The correct answer is 3/4. Try better next time!")
            self.questions -= 1

    def q6(self):
        question = input("What is 5 divided by 10 in fraction?\n")
        solution = "1/2"
        if question == solution:
            self.counter += 1
            print("Current score:", self.counter)
            self.questions -= 1
        else:
            i = 1
            while i < 4:
                print(f"Hint: Remember to simplify! Try again!\nAttempt {i}: ", end="")
                attempt = input()
                if attempt == solution:
                    self.counter += 1
                    print("Current score:", self.counter)
                    self.questions -= 1
                    return
                else:
                    i += 1
            print("The correct answer is 1/2. Try better next time!")
            self.questions -= 1

    def q7(self):
        question = input("If 1% is 0.01, what is 2% in decimal?\n")
        solution = "0.02"
        if question == solution:
            self.counter += 1
            print("Current score:", self.counter)
            self.questions -= 1
        else:
            i = 1
            while i < 4:
                print(f"Hint: Take a closer look at the hundredth place. Try again!\nAttempt {i}: ", end="")
                attempt = input()
                if attempt == solution:
                    self.counter += 1
                    print("Current score:", self.counter)
                    self.questions -= 1
                    return
                else:
                    i += 1
            print("The correct answer is 0.02. Try better next time!")
            self.questions -= 1

    def q8(self):
        question = input("If 20% is 1/5, what is 40% in fraction?\n")
        solution = "2/5"
        if question == solution:
            self.counter += 1
            print("Current score:", self.counter)
            self.questions -= 1
        else:
            i = 1
            while i < 4:
                print(f"Hint: 40% is 20% multiplied by 2. Try again!\nAttempt {i}: ", end="")
                attempt = input()
                if attempt == solution:
                    self.counter += 1
                    print("Current score:", self.counter)
                    self.questions -= 1
                    return
                else:
                    i += 1
            print("The correct answer is 2/5. Try better next time!")
            self.questions -= 1

    def q9(self):
        question = input("If 1/2 x 1/2 is 1/4, what is 2/4 x 2/4?\n")
        solution = "1/4"
        if question == solution:
            self.counter += 1
            print("Current score:", self.counter)
            self.questions -= 1
        else:
            i = 1
            while i < 4:
                print(f"Hint:\nMultiplying fractions mean that you can only multiply the numerator together and the denominator together, respectively. Try again!\nAttempt {i}: ", end="")
                attempt = input()
                if attempt == solution:
                    self.counter += 1
                    print("Current score:", self.counter)
                    self.questions -= 1
                    return
                else:
                    i += 1
            print("The correct answer is 1/4. Try better next time!")
            self.questions -= 1

    def q10(self):
        question = input("If 1/4 x 1/2 is 1/8, what is 1/2 x 1/8?\n")
        solution = "1/16"
        if question == solution:
            self.counter += 1
            print("Current score:", self.counter)
            self.questions -= 1
        else:
            i = 1
            while i < 4:
                print(f"Hint:\nMultiplying fractions mean that you can only multiply the numerator together and the denominator together, respectively. Try again!\nAttempt {i}: ", end="")
                attempt = input()
                if attempt == solution:
                    self.counter += 1
                    print("Current score:", self.counter)
                    self.questions -= 1
                    return
                else:
                    i += 1
            print("The correct answer is 1/16. Try better next time!")
            self.questions -= 1

# method to call all questions
    def qall(self):
        self.q1()
        self.q2()
        self.q3()
        self.q4()
        self.q5()
        self.q6()
        self.q7()
        self.q8()
        self.q9()
        self.q10()




News_Parser/news.py

import requests
import config
from fpdf import FPDF
from argparse import ArgumentParser


def get_news(url, filename, args):

    news = requests.get(url).json()
    # print(news)

    articles = news['articles']
    # print(articles)

    collection = {}
    n = 1
    for arti in articles:
        title = arti['title'].encode('latin-1', 'replace').decode('latin-1')
        source = arti['source']['name']
        link = arti['url']
        collection[n] = title, source, link
        n += 1

    title = args.topic.upper()

    pdf = FPDF()
    pdf.add_page()
    pdf.set_font('times', 'B', size=15)

    pdf.cell(txt=title, new_x="LEFT", new_y="NEXT")
    pdf.cell(txt=" ", new_x="LEFT", new_y="NEXT")
    pdf.set_font('times', size=12)

    for n in range(1, len(articles)+1):
        pdf.set_font('times', 'I', size=15)
        # pdf.set_fill_color(200, 220, 255)
        headline = collection[n][0]
        source_art = collection[n][1]
        URL = collection[n][2]
        if len(headline) > 80:
            headline = headline[:80] + '...'
        pdf.cell(txt=headline, new_x="LEFT", new_y="NEXT", link=URL)
        pdf.set_font('times', 'B', size=10)
        pdf.cell(txt=source_art, new_x="LEFT", new_y="NEXT")
        pdf.cell(txt=" ", new_x="LEFT", new_y="NEXT")

    pdf.output(filename)


def file_ext_check(args):
    filename = args.f
    return filename if filename.endswith('.pdf') else filename + '.pdf'


def link_generator(args):
    if args.topic is not None:
        topic_name = 'q=' + args.topic + '&'

    else:
        topic_name = ''
    if args.country is not None:
        country_name = 'country=' + args.country + '&'
    else:
        country_name = ''
    if args.fm is not None:
        fm = 'from=' + args.fm + '&'
    else:
        fm = ''
    if args.to is not None:
        to = 'to=' + args.to + '&'
    else:
        to = ''
    if args.source is not None:
        src = 'domain=' + args.source + '&'
    else:
        src = ''
    api_key = 'apiKey=' + config.api_key
    url = f'https://newsapi.org/v2/everything?{src}{topic_name}{fm}{to}{country_name}{api_key}'
    return url


def get_args():
    ap = ArgumentParser()
    ap.add_argument('--topic', help='topic name', default=None)
    ap.add_argument('--source', help='source example: wsj.com', default=None)
    ap.add_argument('--fm', help='from yyyy-mm-dd', default=None)
    ap.add_argument('--to', help='to yyyy-mm-dd', default=None)
    ap.add_argument('--country', help='us, sg, uk', default=None)
    ap.add_argument('--f', help='filename', default='news.pdf')
    return ap.parse_args()


def main():
    args = get_args()
    file = file_ext_check(args)
    url = link_generator(args)
    get_news(url, file, args)


if __name__ == '__main__':
    main()



Mathster Project/login_page.py
# To create a Login object
import csv
import sys
import time
from pwinput import pwinput
import validators


class User:
    def email(self):
        self.email = input("Email: ").strip()
        if validate(self.email) != True:
            print("Email found to be invalid.")
            return None
        else:
            return self.email

    def login(self):
        self.username = input("Username: ")
        #self.email = input("Email: ")
        self.password = pwinput()
        user_data = {"username": self.username, "password": self.password}
        with open("userbase.csv") as file:
            reader = csv.DictReader(file, fieldnames=["username", "password"])
            for row in reader:
                if row["username"] == user_data["username"] and row["password"] == user_data["password"]:
                    return ("Success")
        return ("Unsuccessful")


    def register(self):
        # creating a new e-mail function as 2FA for forget password feature
        if self.email() == None:
            return None
           # checking to see if username exists in userbase.csv
        with open("userbase.csv") as file:
            reader = csv.DictReader(file, fieldnames=["username", "password", "email"])
            file.seek(0)
            for row in reader:
                if row["email"] == self.email:
                    print("Email is registered with an account already.")
                    return None
        self.username = input("Username: ").lower().strip()
        # checking to see if username exists in userbase.csv
        with open("userbase.csv") as file:
            reader = csv.DictReader(file, fieldnames=["username", "password"])
            file.seek(0)
            for row in reader:
                if row["username"] == self.username:
                    print("Username already exists")
                    return None
        # requests for confirmation of password
        while True:
            pw = pwinput(prompt="Enter desired password: ")
            print("Password length:", len(pw))
            pw_2 = pwinput(prompt="Confirm password: ")
            if pw != pw_2:
                print("Password did not match")
                continue
            else:
                password = pw
                self.password = password
                break
        # saves username and password into a csv file
        with open("userbase.csv", "a") as file:
            file.seek(0)
            writer = csv.DictWriter(file, fieldnames=["username", "password", "email"])
            writer.writerow({"username": self.username, "password": self.password, "email": self.email})
        ans_question = input("Which Primary School did you go to?\n")
        with open("security.csv", "a") as file:
            file.seek(0)
            writer = csv.DictWriter(file, fieldnames=["answer", "email"])
            writer.writerow({"answer": ans_question.strip(), "email": self.email.strip()})
        return "Success"

    def forget(self):
        self.email = input("Enter the email address you registered with.\n")
        self.username = input("Enter the username you registered with.\n")
        ans_question = input("Security Question: Which Primary School did you go to?\n")
        security_data = {"answer": ans_question, "email": self.email}
        y = '' # just a random variable for control-flow purposes
        with open("security.csv") as file:
            file.seek(0)
            reader = csv.DictReader(file, fieldnames=["answer", "email"])
            for row in reader:
                if row == security_data:
                    y = 'true' # allows method to print forgotten password
        if y != 'true':
            return
        with open("userbase.csv") as file:
            file.seek(0)
            reader = csv.DictReader(file, fieldnames=["username", "password", "email"])
            for row in reader:
                if row["username"] == self.username:
                    print(f"Your password was... {row['password']}")
                    return "Success"

    def homepage(self):
        i = 2
        while i < 3:
            print_slow("Hello, welcome to the login page.\n")
            txt = "1. Login\n2. Register\n3. Forgot Password\n4. Exit Program\n"
            choice = input(txt)
            _choice = ["1", "2", "3", "4"]

            if choice not in _choice:
                i += 1
                print(f"Invalid choice selected. 1-4 only. Try again. {i}/3")

            elif choice == _choice[0]: # Login feature
                if self.login() != "Success":
                    print("Login was unsuccessful")

                else:
                    print(f"{self.username} has logged into the program successfully.")
                    return self.username

            elif choice == _choice[1]: # Registration feature
                if self.register() != "Success":
                    print("Registration was unsuccessful")
                else:
                    print(f"{self.username}'s registration was successful.")


            elif choice == _choice[2]: # Forget password feature
                if self.forget() != "Success":
                    print("Invalid information was inputted.")
                else:
                    print("Password successfully recovered.")
            elif choice == _choice[3]:
                sys.exit("Exiting program now...")
        sys.exit("Quitting")


# typewriter effect
# source: Stackoverflow
def print_slow(str):
    for char in str:
        time.sleep(0.03)
        sys.stdout.write(char)
        sys.stdout.flush()

# for validating email addresses
def validate(s):
    return (validators.email(s))




project/project.py
import login_page
import level1
import time
import csv

def main():
    welcome()
    name = user()
    level_1(name)
    # calls for score & rank
    s = score(name)
    rank = ranker(s)
    next_rank(rank)


# returns score
def score(n: str) -> int:
    with open("scores.csv") as f:
        columns = ["name", "score"]
        reader = csv.DictReader(f, columns)
        data = [row for row in reader if row["name"] == n]
        score = int(data[0]["score"])
        return score
"""
returns rank according to score
>90 - "Math_Elite"
80-89 - "Math_Diamond"
60-79 - "Math_Platinum"
50-59 - "Math_Gold"
40-49 - "Math_Silver+"
30-39 - "Math_Silver"
20-29 - "Math_Bronze+"
10-19 - "Math_Bronze"
"""
def ranker(r: int) -> tuple:
    if r >= 90:
        return "Math_Elite", r
    elif r >= 80:
        return "Math_Diamond", r
    elif r >= 60:
        return "Math_Platinum", r
    elif r >= 50:
        return "Math_Gold", r
    elif r >= 40:
        return "Math_Silver+", r
    elif r >= 30:
        return "Math_Silver", r
    elif r >= 20:
        return "Math_Bronze+", r
    elif r >= 10:
        return "Math_Bronze", r
    elif r >= 0: # eliminate possibility that r is negative
        return "Noob", r
    else:
        raise ValueError("r must be a non-negative integer")

def next_rank(r: tuple) -> str:
    # If tuple index 0 is Noob, look to the right of list and calculate (entry - score)
    # where entry = starting score of next rank
    ranks = ["Noob", "Math_Bronze", "Math_Bronze+", "Math_Silver",
    "Math_Silver+", "Math_Gold", "Math_Platinum", "Math_Diamond", "Math_Elite"
    ]
    entry = [0, 10, 20, 30, 40, 50, 60, 80, 90]
    for _ in ranks:
        i = ranks.index(r[0])
    try:
        # next rank
        n = ranks[i+1]
        # score to next rank
        s = entry[i+1] - int(r[1])
        print(f"You are currently {ranks[i]}. Next rank is {n}. You need to earn {s} more points to rank up!")
        return r
    except IndexError:
        print(f"Maximum rank of Math_Elite obtained... \nYou don't need a rank to keep improving on Mathematics, continue to work hard!;)")
        time.sleep(1.5)
        return r

def welcome():
    txt = "Name of program: Mathster\nCreated for: Primary School Kids\nPurpose: Make Math Fun!"
    d = 33 * "*"
    print(f"{d}\n{txt}\n{d}")
    # time.sleep(1.0)

def user():
    user = login_page.User()
    return user.homepage()

def level_1(n):
    level = level1.Level_One(n)
    print(level)
    level.check_ready()
    level.qall()



if __name__ == "__main__":
    main()

